File: api/int-reztro-restaurant-mgt-dashboard-api-dev/app/utils/cloudinary.py
import cloudinary
import cloudinary.uploader

from fastapi import APIRouter, UploadFile, File, HTTPException
from app.core.config import settings

# CLOUDINARY CONFIG

cloudinary.config(
    cloud_name=settings.CLOUD_NAME,
    api_key=settings.API_KEY,
    api_secret=settings.API_SECRET,
    secure=True
)

# Allowed file types
ALLOWED_TYPES = ["image/jpeg", "image/png", "image/webp"]

# ROUTER
router = APIRouter(
    prefix="/images",
    tags=["Images"]
)


def upload_to_cloudinary(file_or_url, folder: str = "reztro"):
    try:

        if hasattr(file_or_url, "file"):

            if file_or_url.content_type not in ALLOWED_TYPES:
                raise HTTPException(
                    status_code=400,
                    detail="Invalid file type. Only JPEG, PNG, WEBP allowed."
                )

            result = cloudinary.uploader.upload(
                file_or_url.file,
                folder=folder
            )

        elif isinstance(file_or_url, str):

            result = cloudinary.uploader.upload(
                file_or_url,
                folder=folder
            )

        else:
            raise HTTPException(
                status_code=400,
                detail="Invalid input for image upload"
            )

        return {
            "url": result["secure_url"],
            "public_id": result["public_id"]
        }

    except HTTPException:
        raise

    except Exception as e:
        raise HTTPException(
            status_code=500,
            detail=f"Upload failed: {e}"
        )


# DELETE IMAGE FROM CLOUDINARY

def delete_from_cloudinary(public_id: str):
    try:
        result = cloudinary.uploader.destroy(public_id)

        return {
            "result": result
        }

    except Exception as e:
        raise HTTPException(status_code=500, detail=f"Delete failed: {str(e)}")


# FASTAPI ROUTES

@router.post("/upload")
async def upload_image(file: UploadFile = File(...)):
    result = upload_to_cloudinary(file)

    return {
        "message": "Image uploaded successfully",
        "data": result
    }


@router.delete("/delete/{public_id}")
async def delete_image(public_id: str):
    result = delete_from_cloudinary(public_id)

    return {
        "message": "Image deleted successfully",
        "data": result
    }

File: nextjs/myapp/helper/utils/cloudinary.js
export const uploadToCloudinary = async (file, setUploading) => {
  try {
    if (!file) throw new Error("No file selected");

    const allowedTypes = ["image/jpeg", "image/png", "image/jpg"];
    if (!allowedTypes.includes(file.type)) {
      throw new Error("Only JPG/PNG images are allowed");
    }

    if (file.size > 2 * 1024 * 1024) {
      throw new Error("File size should be less than 2MB");
    }

    const formData = new FormData();
    formData.append("file", file);
    formData.append(
      "upload_preset",
      process.env.NEXT_PUBLIC_CLOUDINARY_PRESET
    );

    setUploading(true);

    const res = await fetch(
      `https://api.cloudinary.com/v1_1/${process.env.NEXT_PUBLIC_CLOUDINARY_CLOUD}/image/upload`,
      {
        method: "POST",
        body: formData,
      }
    );

    if (!res.ok) {
      throw new Error("Failed to upload image");
    }

    const data = await res.json();

    if (!data.secure_url) {
      throw new Error("Invalid response from Cloudinary");
    }

    return {
      url: data.secure_url,
      public_id: data.public_id, 
    };
  } catch (error) {
    console.error("Cloudinary Error:", error.message);
    throw error;
  } finally {
    setUploading(false);
  }
};